seed/seed20.go

```diff
@@ -137,47 +186,279 @@ func (s *seed20) Copy(seedDir string, label string, tm timings.Measurer) (err er
-		if err != nil {
+	dirs := []string{"snaps", "assertions"}
+	for _, d := range dirs {
+		if err := os.Mkdir(filepath.Join(destSystemDir, d), 0755); err != nil {
 			return err
 		}
+	}
 
-		destPath := filepath.Join(destSeedDir, "systems", label, strings.TrimPrefix(path, srcSystemDir))
-		if info.IsDir() {
-			return os.Mkdir(destPath, info.Mode())
-		}
-
-		return osutil.CopyFile(path, destPath, osutil.CopyFlagDefault)
-	})
-	if err != nil {
+	if err := osutil.CopyFile(
+		filepath.Join(srcSystemDir, "model"),
+		filepath.Join(destSystemDir, "model"),
+		osutil.CopyFlagDefault,
+	); err != nil {
 		return err
 	}
 
 	destAssertedSnapDir := filepath.Join(destSeedDir, "snaps")
-
-	if err := os.MkdirAll(destAssertedSnapDir, 0755); err != nil {
+	if err := os.Mkdir(destAssertedSnapDir, 0755); err != nil {
 		return err
 	}
 
-	// copy the asserted snaps that the seed needs
+	destUnassertedSnapDir := filepath.Join(destSystemDir, "snaps")
+
+	// while copying the snaps, we also collect the assertions and the fields in
+	// the aux-info.json file that we need to write
+	var assertions []asserts.Assertion
+	auxInfo := make(map[string]*internal.AuxInfo20)
+	optSnaps := make([]*internal.Snap20, 0)
+
+	// copy the snaps and components that the seed needs
 	for _, sn := range s.snaps {
-		// unasserted snaps are already copied above, skip them
-		if sn.ID() == "" {
+		// if we're not copying the snap, then we also don't need to copy the
+		// components for this snap
+		if !shouldCopySnap(sn, s.model, s.modelSnaps, opts.OptionalContainers) {
 			continue
 		}
 
-		destSnapPath := filepath.Join(destAssertedSnapDir, filepath.Base(sn.Path))
+		// optSnap might be nil if it shouldn't have an entry in options.yaml
+		as, optSnap, err := s.copySnapAndComponents(sn, destSeedDir, opts)
+		if err != nil {
+			return err
+		}
+
+		if optSnap != nil {
+			optSnaps = append(optSnaps, optSnap)
+		}
+
+		assertions = append(assertions, as...)
+		if sn.ID() != "" {
+			if a, ok := s.auxInfos[sn.ID()]; ok {
+				auxInfo[sn.ID()] = a
+			}
+		}
+	}
+
+	if len(optSnaps) > 0 {
+		opts := internal.Options20{Snaps: optSnaps}
+		if err := opts.Write(filepath.Join(destSystemDir, "options.yaml")); err != nil {
+			return err
+		}
+	}
+
+	if err := writeAuxInfo(filepath.Join(destUnassertedSnapDir, "aux-info.json"), auxInfo); err != nil {
+		return err
+	}
+
+	// copy the assertions that the seed needs. since we don't make any
+	// distinction between the assertions in "extra-snaps" and "snaps" files, we
+	// just write them all to the same "snaps" file.
+	if err := resolveAndSaveAssertions(assertions, s.db, filepath.Join(destSystemDir, "assertions", "snaps")); err != nil {
+		return err
+	}
 
-		if err := osutil.CopyFile(sn.Path, destSnapPath, osutil.CopyFlagOverwrite); err != nil {
-			return fmt.Errorf("cannot copy asserted snap: %w", err)
+	for _, name := range []string{"assertions/model-etc", "grubenv"} {
```

Comment:
The model could be copied in this loop too. Also I worry a bit that we might forget in the future to update this list in case we start to drop more files in the seed. Possibly something to check for in a future spread test (that there are no unexpected files in the original seed).

Reply:
I agree with Alfonso about having a spread test that looks at the diff between origin and copy, both for a full copy and some optional left out copy